### ISSUE TITLE
[1880] Rules Implementation

### ISSUE THREAD

**Issue by bentziaxl** [tobymao/18xx#8631]
[rules pdf ](https://lookout-spiele.de/upload/en_1880china.html_1880_Regeln_115_EN_WEB.pdf)

### Capital
- Supports 3 - 7 players, receiving {600,480,400,340,300} 
- Bank is unlimited, rules say 37860 in paper money

### Auction
- [x] Auction one Private at a time
- [x] no reentry
- [x] all passed on P0 and P1, reduce price by 5, restart auction.
- [x] if goes down to 0, forced win
- [x] otherwise mini OR, privates pay
- [x] keep going until all privates have been sold
- [x] reorder players min cash to max 

### Draft
- 7 foreign investors
- [x] players pick one investor each, no cost
- [x] draft is done once each player picked an investor 

### Stock Rounds
- Sell any number of shares, then buy up to one share, or pass
- Priority order works as normal (left of the last player to take an action in SR)

#### Launching Corporations
- [x] 20% float at initial SR, increases in later phases
- [x] Only president shares count towards floating requirements.
- [x] If a corporation is sold out corp floats, even if president doesn't have enough shares to float (sold out excluding 50% shares reserved at the start of game, and the foreign investors's share)
- [x] When launching can decide 20%, 30%, 40% president share
- [x] After float pick phase permit (single phase for 40%, {AB, BC, CD} for 30%, {ABC, BCD} for 20%)
- [x] Company gets 50% capitalization
- [x] Tokens on map are placed at end of SR when company floats
- [x] pick one of 4 spots for each Par price. This determines corporation operation order for the rest of the game
- [x] If launched corporation is a player's first launch corp, one share is reserved to the foreign investor
- [x] Float chart allowing players to pick par and operating order spot
- [x] ABC/BCD choice for launched BCR at auction
- [ ] Need an easier way to tell what percentage the president share is in the UI. 

#### Share Transactions
- [x] Shares may not be sold in the first Share Round (handled by `SELL_AFTER = :first`)
- [x] each 10% sold has a 5Y fee
- [x] until phase B1, only 50% is available, rest of the shares are blocked
- [x] Can own up to 100% of a corporation
- [x] No limit to shares in market
- [x] Shares sold as the result of an emergency financing may not trigger a change of Director
- [x] When entering phase B1, all companies with 50% shares sold (including FI) get the other 50% capitalization. The 5 remaining shares are available for purchase.
- [x] After phase B1, once the 50% shares sold (including F1), the corporation gets the other 50% capitalization 

#### End of Stock Round Movement
- [x] one price drop per 10% sold 
- [x] share price moves up if all available shares are sold (less then 100% in phases B1, and while Foreign investor owns a share

#### Round Order
- [x] Keep track of last purchased train, if a full set of corporations pass on buying a train type that type is removed from the game
- [x] Upon purchase of last train type (or removal), a SR happens in the middle of a corporations OR - buy train step
- [x] After SR completion, corporation resumes at train purchase step of the OR
- [x] Do not pay privates when coming back to OR
- [x] handle new corporations added during the SR
- [x] handle OR/SR name 
- [x] do not export all trains if an SR occurred during a corps OR, and they refuse to buy the new train

### Operating Rounds
- Privates Pay
- Foreign Investors operate in order
- Corporation operate

### Foreign Investors
- [x] lease current available train from the bank at no cost. 
- [x] can only build or upgrade one tile 
- [x] Income from running trains is witheld
- [x] has one share of owner's first launched corporation. That share doesn't pay dividends when owned by FI
- [x]  When FI and associated corporations are connected (ignoring tokens), Inventor is merged into the corporation
- [x] Owner of the FI chooses 100% or 20% of treasury of the FI to move to Corporation treasury
- [x] Share of corporation is transferred to Owner + Y50
- [x] Corporation may replace FI token with one of its own for free
- [x] automatically discard FI token when corp has no free tokens left
- [x] FI do not operate after phase B3
- [x] if not merged by end of game, 20% of treasury paid to owner

#### Laying Track
- [x] Can only lay track if corp has valid building permit for the phase
- [x] Phase A, can only lay 1 yellow tile (BCR can always lay two), Phase B can lay two yellow or upgrade one tile
- [x] some tiles have 1 town/city or 2 town/city, those can be upgraded to either single city/town or double city/town
- [x] Fix double city/town starting hexes (N4, H8) not rendering the second town
- [x] when upgrading brown city, must use 6 exit tile. (Tile 611, which just has 5 exits, is reserved for locations on the edge of the playing area. Only if all the tiles numbered 63 have been placed, it can be used instead of tile 63.)
- [x] Verify Beijing upgrades from Green to Brown correctly

#### Placing Tokens
- [x] in phase D token cost is doubled
- [x] home tokens are reserved. For the two corporations starting in OO tiles, no token can be placed in either slots until the corporation has floated and picked which city to place its home token. If upgraded to green, a token can be placed as long as there's space there for the reserved corporation

#### Running Trains / Dividends
- [x] restored 2 trains. available after 6, do not rust, do not fulfill train owning obligation, occupy train slot, do not change phase when bought out, do not count as train purchase for SR purposes
- [x] Share price bonus is added to the run of +5/+10/+20/+40 per share depending on corporation share location in the market
- [x] Y50 bonus for connecting Russia and validvostok
- [x] Taiwan company bonus
- [x] full pay (share price moves right) or withold (share price moves left)

#### Communism
- After purchase of 4 train, until purchase of first 6 train
- [x] Share prices are frozen, no changes for any reason
- [x] Director can not sell shares

#### Purchasing Trains
- [x] must own train at end of OR.
- [x] can cross buy if both corporations owned by the same director, only after 3-train is bought
- [x] when bought from the bank, train marker is placed near the company. If the entire set of operating corporations passed on buying trains including the corporation with the marker, the entire train type is removed
- [x] After last 8 train is bought, end game is triggered, 3 more ORs, train marker isn't moved
- [x] Company can pass buying trains triggering an SR, then buy the new train when it's back to OR
- [x] 10 trains and restored 2s are never removed
- [x] Trains removed due to train limit changes are discarded

#### Routes
- [x] can start and end a route in beijing in green & yellow
- [x] red off-board counts as a city
- [x] blue off-board counts as town
- [x] using the ferry connections is -10 to the run

#### Forced Train Purchase/EMR
- [x] train can only be bought from bank
- [x] Owner pays from pocket, sells shares
- [x] No change of ownership during EMR
- [x] No bankruptcy, player is in debt
- [x] 50% is added to the debt immediately at the end of EMR, and at the end of  every SR.
- [x] Player in debt may pay off part or all during SR, can't buy shares until debt is paid off.

#### Private Companies
- [x] P2 applies to every corporation & FI owned by owner
- [x] P3 applies to every corporation & FI owned by owner
- [x] P4  applies to every corporation & FI owned by owner
- [x] P0 owner can choose when to claim the payment (40/70/100 when 2+2/3/3+3 train is bought/removed), if unclaimed 100 is awarded automatically upon 3+3 purchase
- [x] P5 assigns a D permit at any point. Can not change assignments, can't be assigned to FI
- [x] P6, par price is set at 100, owner gets 20% president certificate. BCR immediately floats 
- [x] P7 can be exchanged into train at any point during OR. 
- [x] Purchase train ability shows cost of train instead of free. UI bug
- [x] train purchase marker is not advanced. 
- [x] Forced exchange at phase B3.
- [x] If none of the corporations owned by the owner of P7 have a free train slot, a train can be removed
- [x] if owner of P7 doesn't own a corporation, P7 is discarded.

### Game End Triggers
- after sale of last 8 train (or it's removed), there are 3 more ORs ending with company with the train purchase marker

**Comment by bentziaxl** [comment]
for P4 i had to change corp.all_abilities to return also abilities by corp owning player that are "Owning player, etc"
This breaks a few games.
1889, 18gb -> I created a local corporation.rb class with the old method definition
18fl -> changed the ability when to player_owning_track

Those are workaround, need to submit a different PR for those games, with some pinning involved

**Comment by bentziaxl** [comment]
bugs:
Some private abilities show up for privates (assign D to private, rocket of china for private)
Re-check priority deal
